packages/test/src/test-integration-update.ts

```diff
@@ -0,0 +1,377 @@
+    arg[0] = 'mutated!';
+  };
+  wf.setHandler(updateWithMutableArg, updateHandler, { validator });
+  await wf.condition(() => false);
+}
+
+test('Update handler does not see mutations to arguments made by validator', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const worker = await createWorker();
+  await worker.runUntil(async () => {
+    const wfHandle = await startWorkflow(workflowWithMutatingValidator);
+    const updateResult = await wfHandle.executeUpdate(updateWithMutableArg, { args: [['1']] });
+    t.deepEqual(updateResult, ['1']);
+  });
+});
+
+// The following tests test dispatch of buffered updates. An update is pushed to
+// the buffer if its handler is not available when attempting to handle the
+// update. If the handler is subsequently set by a setHandler call during
+// processing of the same activation, then the handler is invoked on the
+// buffered update . Otherwise, the buffered update is rejected. Hence in order
+// to test dispatch of buffered updates, we need to cause the update job to be
+// packaged together with another job that will cause the handler to be set
+// (e.g. startWorkflow, or completeActivity). This scenario is typically
+// encountered in the first WFT, and that is what these tests recreate. They
+// start the workflow with startDelay, and then send an update (without waiting
+// for the server's response) to ensure that doUpdate and startWorkflow are
+// packaged in the same WFT (despite the large startDelay value, the server will
+// dispatch a WFT when the update is received).
+
+const stateMutatingUpdate = wf.defineUpdate('stateMutatingUpdate');
+
+export async function setUpdateHandlerAndExit(): Promise<string> {
+  let state = 'initial';
+  const mutateState = () => void (state = 'mutated-by-update');
+  wf.setHandler(stateMutatingUpdate, mutateState);
+  // If an Update is present in the first WFT, then the handler should be called
+  // before the workflow exits and the workflow return value should reflect its
+  // side effects.
+  return state;
+}
+
+test('Update is always delivered', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const wfHandle = await startWorkflow(setUpdateHandlerAndExit, { startDelay: '10000 days' });
+
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(stateMutatingUpdate);
+  const worker = await createWorker();
+  await worker.runUntil(async () => {
+    // Worker receives activation: [doUpdate, startWorkflow]
+    const wfResult = await wfHandle.result();
+    t.deepEqual(wfResult, 'mutated-by-update');
+  });
+});
+
+test('Two Updates in first WFT', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const wfHandle = await startWorkflow(workflowWithUpdates, { startDelay: '10000 days' });
+
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(update, { args: ['1'] });
```

Comment:
Don't leave any floating promises, not even in tests. In recent Node versions, unhandled promise failures will, by default, causes node to crashes. `ava` probably catches these, but we should not rely on that.